chain/checkpointing/sub.go

```diff
@@ -0,0 +1,506 @@
+
+		ids := c.formIDSlice(idsStrings)
+
+		id := party.ID(c.host.ID().String())
+
+		threshold := 2
+		n := NewNetwork(ids, c.sub, c.topic)
+		f := frost.KeygenTaproot(id, ids, threshold)
+
+		handler, err := protocol.NewMultiHandler(f, []byte{1, 2, 3})
+		if err != nil {
+			fmt.Println(err)
+			panic(err)
+		}
+		c.LoopHandler(ctx, handler, n)
+		r, err := handler.Result()
+		if err != nil {
+			fmt.Println(err)
+			panic(err)
+		}
+		fmt.Println("Result :", r)
+
+		c.newconfig = r.(*keygen.TaprootConfig)
+
+		return true, nil
+	}
+
+	revertHandler := func(ctx context.Context, ts *types.TipSet) error {
+		return nil
+	}
+
+	match := func(oldTs, newTs *types.TipSet) (bool, events.StateChange, error) {
+		newAct, err := c.api.StateGetActor(ctx, mpower.PowerActorAddr, newTs.Key())
+		if err != nil {
+			return false, nil, err
+		}
+
+		oldAct, err := c.api.StateGetActor(ctx, mpower.PowerActorAddr, oldTs.Key())
+		if err != nil {
+			return false, nil, err
+		}
+
+		var oldSt, newSt mpower.State
+
+		bs := blockstore.NewAPIBlockstore(c.api)
+		cst := cbor.NewCborStore(bs)
+		if err := cst.Get(ctx, oldAct.Head, &oldSt); err != nil {
+			return false, nil, err
+		}
+		if err := cst.Get(ctx, newAct.Head, &newSt); err != nil {
+			return false, nil, err
+		}
+
+		// If Power Actors list has changed start DKG
+		if !c.init {
+			ts, err := c.api.ChainGetTipSetByHeight(ctx, 0, oldTs.Key())
+			if err != nil {
+				panic(err)
+			}
+			data := ts.Cids()[0]
+			err = c.initiate(data.Bytes())
+			if err != nil {
+				panic(err)
+			}
+			return false, nil, nil
+		}
+
+		// ZONDAX TODO
+		// Activate checkpointing every 20 blocks
+		fmt.Println("Height:", newTs.Height())
+		if newTs.Height()%20 == 0 {
+			fmt.Println("Check point time")
+
+			// Initiation and config should be happening at start
+			if c.init && c.config != nil {
+
+				data := oldTs.Cids()[0]
+
+				var partyList string = ""
+				for _, partyId := range c.orderParticipantsList() {
+					partyList += partyId + "\n"
+				}
+				hash, err := CreateConfig([]byte(partyList))
+				if err != nil {
+					panic(err)
+				}
+
+				c.CreateCheckpoint(ctx, data.Bytes(), hash)
+			}
+		}
+
+		// Changes detected so generate new key
+		if oldSt.MinerCount != newSt.MinerCount {
+			fmt.Println("Generate new config")
+
+			return true, nil, nil
+		}
+
+		return false, nil, nil
+	}
+
+	err := c.events.StateChanged(checkFunc, changeHandler, revertHandler, 5, 76587687658765876, match)
+	if err != nil {
+		return
+	}
+}
+
+func (c *CheckpointingSub) Start(ctx context.Context) {
+	topic, err := c.pubsub.Join("keygen")
+	if err != nil {
+		panic(err)
+	}
+	c.topic = topic
+
+	// and subscribe to it
+	sub, err := topic.Subscribe()
+	if err != nil {
+		panic(err)
+	}
+	c.sub = sub
+
+	c.listenCheckpointEvents(ctx)
+}
+
+func (c *CheckpointingSub) LoopHandler(ctx context.Context, h protocol.Handler, network *Network) {
+	for {
+		msg, ok := <-h.Listen()
+		if !ok {
+			network.Done()
```

Comment:
If you don't need to clean anything, you can probably remove this function and just return when the channel is closed.